### PR TITLE
Add credential provider support

### PR DESCRIPTION
**Pull request by trentbitterman**
Adds an abstract credential provider class that can be used to programmatically supply credentials at
authentication time, instead of just a static username and password.

### CODE REVIEW

**Comment by trentbitterman** [comment]
Closes #246

**Review by alisaifee — CHANGES_REQUESTED** [review]
Generally this looks great! Thank you for the comprehensive and very clean contribution!

I've left a few minor suggestions for your consideration. Additionally if you have an implementation for an IAM provider for example - perhaps this could be added to `coredis.recipes` and referenced in the documentation?

File: coredis/credentials.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    ) -> None:
+        self.username = username or ""
+        self.password = password or ""
+
+    def get_credentials(self) -> PassOnly | UserPass:
+        return (
+            PassOnly(self.password)
+            if not self.username
+            else UserPass(self.username, self.password)
+        )
```
[the remark this answers is not in the document]
Comment (trentbitterman) [reply coredis/credentials.py:45]: I like that, although I think we would need to modify the return type of the abstract method too if we were going to simplify the code in the connection class. Did you want that changed as well?

[the remark this answers is not in the document]
Comment (alisaifee) [reply coredis/credentials.py:45]: Ah yes, missed that tweak!

**Comment by trentbitterman** [comment]
@alisaifee Thanks for the review! Let me know if there are any other changes you'd like me to make.

**Comment by alisaifee** [comment]
> @alisaifee Thanks for the review! Let me know if there are any other changes you'd like me to make.

Wow, amazing turnaround to address the suggestion! Thank you! Looks great to me. I wasn't able to verify the rendering of the docs since I've broken my local sphinx environment but that's a minor detail we can iron out in master.

Regarding the typing issues in CI with respect to boto - feel free to add it to the ignore list.

**Comment by trentbitterman** [comment]
@alisaifee Hopefully that's the correct place to ignore them. Let me know if it's not.

File: pyproject.toml [hunk 2]
```diff
@@ -29,6 +29,8 @@ warn_unused_ignores = true
 module = [
     "async_timeout",
     "beartype",
+    "botocore",
```
Comment (alisaifee) [inline pyproject.toml:32]:
```suggestion
    "botocore.*",
```

File: coredis/recipes/credentials/iam_provider.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        self.request_signer = RequestSigner(
+            ServiceId("elasticache"),
+            self.region,
+            "elasticache",
+            "v4",
+            session.get_credentials(),
+            session.get_component("event_emitter"),
+        )
+
+    @cached(cache=TTLCache(maxsize=128, ttl=900))
```
Comment (alisaifee) [inline coredis/recipes/credentials/iam_provider.py:36]:
```suggestion
    @cached(cache=TTLCache(maxsize=128, ttl=900)). # type: ignore[misc]
```

**Comment by alisaifee** [comment]
> @alisaifee Hopefully that's the correct place to ignore them. Let me know if it's not.

It is indeed. Needed a few tweaks to get past mypy. Since it's a recipe it's fine to be a bit relaxed on the type strictness IMO.

**Review by alisaifee — APPROVED** [review]

**Comment by alisaifee** [comment]
@trentbitterman unless you have anything else to add, I'll probably merge this once CI is done.

**Comment by trentbitterman** [comment]
That works for me, I didn’t have any other changes planned. Thanks!

**Pull request merged**